File: 1669-minimum-cost-to-cut-a-stick/minimum-cost-to-cut-a-stick.cpp
class Solution {
public:
    int minCost(int n, vector<int>& cuts) {
        cuts.insert(cuts.begin(), 0);
        cuts.push_back(n);
        sort(cuts.begin(), cuts.end());
        int c = cuts.size();

        vector<vector<int>> dp(c, vector<int>(c, 0));
        for (int diff = 2; diff < c; diff++) {
            for (int l = 0; l < c - diff; l++) {
                int r = l + diff;
                // cout << l << " " << r << endl;
                int ans = INT_MAX;
                for (int i = l + 1; i < r; i++) {
                    ans = min(ans, cuts[r] - cuts[l] + dp[l][i] + dp[i][r]);
                }
                dp[l][r] = ans;
            }
        }

        return dp[0][c - 1];
    }
};

File: 1669-minimum-cost-to-cut-a-stick/minimum-cost-to-cut-a-stick.py
class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        cuts.sort()
        cuts = [0] + cuts + [n]

        @cache
        def func(l, r):
            if (abs(r - l) <= 1):
                return 0
                
            res = float('inf')
            for i in range(l + 1, r):
                res = min(res, cuts[r] - cuts[l] + func(l, i) + func(i, r))
                print(res)
            return res

        return func(0, len(cuts) - 1)